[MRG] Affinity prop fix

absolutelyNoWarranty opened the pull request:
Commit 1. Changes defaults to address issues in #5340
Commit 2. Issue warning if affinity propagation doesn't converge; error if fit is called after.

absolutelyNoWarranty committed:
Change default parameters of affinity propagation

See https://github.com/scikit-learn/scikit-learn/issues/5340
Changes damping to 0.9 for better clustering results.
Uses Euclidean distances instead of squared Euclidean.

Default maximum number of iterations left unchanged.

absolutelyNoWarranty committed:
Add warning and error checking.

If affinity_clustering does not converge there will be zero
exemplars which we return as an empty list and warn the user
of. Also we prevent the user from calling predict when fit
did not converge.

ogrisel commented on sklearn/cluster/affinity_propagation_.py:
```diff
@@ -288,17 +295,21 @@ def fit(self, X, y=None):
-                convergence_iter=self.convergence_iter, damping=self.damping,
-                copy=self.copy, verbose=self.verbose, return_n_iter=True)
+        with warnings.catch_warnings():
+            warnings.simplefilter('always', ConvergenceWarning)
```
Why is this necessary here? I would rather not mess with the warnings machine in our fit code but rather only the tests.

amueller replied:
+1

ogrisel commented on sklearn/cluster/tests/test_affinity_propagation.py:
```diff
@@ -66,7 +67,14 @@ def test_affinity_propagation_predict():
+    af = AffinityPropagation(affinity="euclidean", max_iter=1)
+    assert_warns(ConvergenceWarning, af.fit, X)
+    assert_raises(ValueError, af.predict, X)
+    
```
Could you please add a non-regression test that checks that a bad damping level can cause the convergence warning?

absolutelyNoWarranty replied:
Is it ok if it requires "magic" parameters? It seems you have to have a particular combination of bad damping and bad preferences combined with the right number of max_iter to get it to not converge. Sometimes it will "converge" but return # clusters = # data ...

amueller replied:
well the test is deterministic. having entirely magic values is probably not that helpful. If the behavior can be produced with a range of values, picking something in that range seems fine to me.

absolutelyNoWarranty replied:
For example, low damping over a range of (outrageous low) preference values: 

	import numpy as np
	from sklearn.cluster import AffinityPropagation
	from sklearn.datasets.samples_generator import make_blobs
	from sklearn.metrics import euclidean_distances

	centers = np.array([[1, 1], [-1, -1], [1, -1]]) + 10
	X, _ = make_blobs(n_samples=60, n_features=2, centers=centers,
					  cluster_std=0.4, shuffle=True, random_state=0)

	S_sq = euclidean_distances(X, squared=True)
	S_nonsq = euclidean_distances(X, squared=False)

	max_sq = int(np.max(S_sq))
	max_nonsq = int(np.max(S_nonsq))

	# low damping over range of preference values

	for p in range(max_nonsq*10, max_nonsq*10+50):
		af = AffinityPropagation(preference=-p, max_iter=1000,
								 convergence_iter=100, damping=0.5,
								 affinity='precomputed')
		af.fit(-S_nonsq)
		n_clusters = len(af.cluster_centers_indices_) if af.cluster_centers_indices_ is not None else 0
		n_iter = af.n_iter_
		
		print 'p=%d: %d after %d iter' % (-p, n_clusters, n_iter)

	for p in range(max_sq*100, max_sq*100+50):
		af = AffinityPropagation(preference=-p, max_iter=1000,
								 convergence_iter=100, damping=0.5,
								 affinity='precomputed')
		af.fit(-S_sq)
		n_clusters = len(af.cluster_centers_indices_) if af.cluster_centers_indices_ is not None else 0
		n_iter = af.n_iter_
		
		print 'p=%d: %d after %d iter' % (-p, n_clusters, n_iter)

and you will get something like 

>p=-40: 2 after 231 iter
>p=-41: 1 after 1000 iter
>p=-42: 2 after 540 iter
>p=-43: 1 after 987 iter
>p=-44: 1 after 561 iter
>p=-45: 1 after 450 iter
>p=-46: 1 after 822 iter
>p=-47: 1 after 517 iter
>p=-48: 1 after 353 iter
>p=-49: 60 after 1000 iter
>p=-50: 2 after 337 iter
>p=-51: 0 after 1000 iter
>p=-52: 60 after 1000 iter
>p=-53: 60 after 1000 iter
>p=-54: 2 after 1000 iter
>p=-55: 1 after 988 iter
>p=-56: 60 after 1000 iter
>p=-57: 60 after 1000 iter
>p=-58: 59 after 1000 iter
>p=-59: 0 after 1000 iter
>p=-2000: 0 after 1000 iter
>p=-2001: 60 after 1000 iter
>p=-2002: 60 after 1000 iter
>p=-2003: 60 after 1000 iter
>p=-2004: 0 after 1000 iter
>p=-2005: 60 after 1000 iter
>p=-2006: 0 after 1000 iter
>p=-2007: 60 after 1000 iter
>p=-2008: 0 after 1000 iter
>p=-2009: 60 after 1000 iter
>p=-2010: 60 after 1000 iter
>p=-2011: 0 after 1000 iter
>p=-2012: 0 after 1000 iter
>p=-2013: 60 after 1000 iter
>p=-2014: 60 after 1000 iter
>p=-2015: 0 after 1000 iter
>p=-2016: 60 after 1000 iter
>p=-2017: 0 after 1000 iter
>p=-2018: 60 after 1000 iter
>p=-2019: 0 after 1000 iter

I suppose a test would look for IF # clusters = 0, THEN should have raised a ConvergenceWarning but does it make sense to let AP get away with return "60 clusters"  (clearly not converging as well ...)
Perhaps the real issue is that sklearn should include some sort of [preference range checking](http://www.psi.toronto.edu/affinitypropagation/preferenceRange.m)?

amueller replied:
On 10/19/2015 06:32 AM, absolutelyNoWarranty wrote:
> Perhaps the real issue is that sklearn should include some sort of 
> preference range checking 
> <http://www.psi.toronto.edu/affinitypropagation/preferenceRange.m>? 
Sounds like a good idea.

amueller commented:
Changing behavior like this is not good. So the reference implements both squared and non-squared distances? Or was that an error in the code?
For the damping at least, I'm afraid we need to do a deprecation, i.e. set the parameter to None, use the old value, and if it is set to None tell people it'll change.

absolutelyNoWarranty commented:
Sorry for being unclear; the reference just takes S, and when given the same S, sklearn and the reference run the same. In my tests the reason why the reference was doing so well (i.e. without getting 4 clusters every now and then is because I was giving it non-squared distances.)

absolutelyNoWarranty committed:
Change defaults back but add DeprecationWarning.

Using default damping=0.5 raises DeprecationWarning
Rename original default affinity parameter from 'euclidean' to
'sqeuclidean', raise DeprecationWarning when 'euclidean' is used

absolutelyNoWarranty committed:
Remove warning filter from Fit

amueller commented on sklearn/cluster/affinity_propagation_.py:
```diff
@@ -288,17 +295,21 @@ def fit(self, X, y=None):
             self.affinity_matrix_ = X
         elif self.affinity == "euclidean":
-            self.affinity_matrix_ = -euclidean_distances(X, squared=True)
+            self.affinity_matrix_ = -euclidean_distances(X, squared=False)
```
I was referring to this change. If this makes the results better, we should change the behavior. But we can't  change is this abruptly.

absolutelyNoWarranty replied:
Add the option for a ``sqeuclidean`` as an option for affinity, but keep ``euclidean`` as default with a deprecated warning that says "in the future please use sqeuclidean for squared distances"?

amueller replied:
yes.

absolutelyNoWarranty committed:
Add additional ConvergenceWarning

Breaks possible ConvergenceWarnings into two 1) if AP has not satisfied
convergence_iter after max_iter is exhausted and 2) if no exemplers were
found.

absolutelyNoWarranty commented:
So I made the changes requested but I have a question about licensing. I would like to include preference range checking but I don't know if that violates their licence/copyright.  Infact, the current affinity_code code in scikit learn seems to be derived from the reference implementation but there's no mention of it.


On the AP site:
>Can I use affinity propagation for a commercial application? If you want to use it to obtain results for commercial purposes, you should contact us for a site license. If you want to incorporate it into software that is to be used for commercial purposes, you should contact us for licensing.

agramfort commented on sklearn/cluster/affinity_propagation_.py:
```diff
@@ -288,6 +302,12 @@ def fit(self, X, y=None):
+                          "sqeuclidean for squared and euclidean for "
+                          "for non-squared distances.",
+                          DeprecationWarning)
+            self.affinity_matrix_ = -euclidean_distances(X, squared=False)
```
this is not the right fix. You'll get a warning all the time even if you know what you're doing ie. want actually "euclidean"

this is also not backward compatible.

absolutelyNoWarranty replied:
Yes it a mistake -- was supposed to be identical to the option below (both would be squared=True).

absolutelyNoWarranty committed:
Fix to match warning message.

absolutelyNoWarranty committed:
Fix typo

amueller commented:
It looks like the tests don't pass on appveyor (don't worry about circleci) :-/

amueller commented:
@GaelVaroquaux @agramfort can you comment on the licensing issue. You are listed as the authors in the file.
Is preference range checking described in the paper? Then there wouldn't be a problem.

agramfort commented:
I wrote the initial version but it was a while ago. I don't remember "copying" the "official" version.

GaelVaroquaux commented:
> I don't remember "copying" the "official" version.

No, we certainly did not do that.

absolutelyNoWarranty commented:
I didn't mean to accuse anyone of anything, it's just that a lot of the comments were the same so I thought maybe the [MATLAB version](http://www.psi.toronto.edu/affinitypropagation/software/apcluster.m) had been used as a template/reference. Apologies for the mistaken assumption. 

(A cursory glance at the two would probably fool most people.)

e.g.
sklearn
```python
I = np.where(np.diag(A + R) > 0)[0]
K = I.size  # Identify exemplars
if K > 0:
        c = np.argmax(S[:, I], axis=1)
        c[I] = np.arange(K)  # Identify clusters
        # Refine the final set of exemplars and clusters and return results
        for k in range(K):
            ii = np.where(c == k)[0]
            j = np.argmax(np.sum(S[ii[:, np.newaxis], ii], axis=0))
            I[k] = ii[j]

        c = np.argmax(S[:, I], axis=1)
        c[I] = np.arange(K)
        labels = I[c]
        # Reduce labels to a sorted, gapless, list
        cluster_centers_indices = np.unique(labels)
        labels = np.searchsorted(cluster_centers_indices, labels)
    else:
        labels = np.empty((n_points, 1))
        cluster_centers_indices = None
        labels.fill(np.nan)
```
 theirs
```MATLAB
I=find((diag(A)+diag(R))>0); K=length(I); % Identify exemplars
if K>0
    [tmp c]=max(S(:,I),[],2); c(I)=1:K; % Identify clusters
    % Refine the final set of exemplars and clusters and return results
    for k=1:K ii=find(c==k); [y j]=max(sum(S(ii,ii),1)); I(k)=ii(j(1)); end; notI=reshape(setdiff(1:N,I),[],1);
    [tmp c]=max(S(:,I),[],2); c(I)=1:K; tmpidx=I(c);
    tmpdpsim=sum(S(sub2ind([N N],notI,tmpidx(notI))));
    tmpexpref=sum(dS(I));
    tmpnetsim=tmpdpsim+tmpexpref;
else
    tmpidx=nan*ones(N,1); tmpnetsim=nan; tmpexpref=nan;
end;
```

agramfort commented:
Our code is a reimplementation of the method starting from the notations in
the paper.
We use different computation primitives but naturally follow the steps of
the algorithms (responsibilities, availabilities, ...). Furthermore our
code has some difference as it was pointed out for example in
https://github.com/scikit-learn/scikit-learn/issues/5340#issuecomment-148323496

However to make sure sklearn users get comparable results as with matlab,
which is crucial for fair benchmarks in academic papers, the code was
modified especially in the code block you pasted above. Note that we have
often had complaints that our implementation do not agree with matlab or R
packages.

absolutelyNoWarranty commented:
@amueller It's not described in the paper unfortunately.

cmarmo commented:
I'm closing this pull request as the related issue is no longer present in master.
